fs/fuse/compound.c

```diff
@@ -0,0 +1,441 @@
+	hdr->unique = fuse_get_unique(&compound->fm->fc->iq);
+	compound->buffer_pos += sizeof(*hdr);
+
+	/* Copy operation arguments */
+	for (i = 0; i < args->in_numargs; i++) {
+		/* Last argument with in_pages flag: copy from pages */
+		if (i == args->in_numargs - 1 && args->in_pages) {
+			/* we have external payload,
+			 * this is not supported at the moment */
+			return -EINVAL;
+		} else {
+			/* Regular argument: copy from value pointer */
+			memcpy(compound->buffer + compound->buffer_pos,
+			       args->in_args[i].value, args->in_args[i].size);
+			compound->buffer_pos += args->in_args[i].size;
+		}
+	}
+
+	/* Track expected output size */
+	compound->expected_out_sizes[compound->count] =
+		sizeof(struct fuse_out_header) + expected_out_size;
+	compound->total_expected_out_size +=
+		compound->expected_out_sizes[compound->count];
+
+	/* Store args for response parsing */
+	compound->op_args[compound->count] = args;
+
+	compound->count++;
+	compound->total_size += needed_size;
+
+	return 0;
+}
+
+/*
+ * Parse compound response
+ *
+ * Parses the compound response and populates the original
+ * fuse_args structures with the response data. This function is idempotent
+ * and can be called multiple times safely.
+ *
+ * For operations with page-based output (out_pages=true), the response data
+ * is ignored at the moment.
+ *
+ * Returns 0 on success, negative error code on failure.
+ */
+static int fuse_compound_parse_resp(struct fuse_compound_req *compound)
+{
+	struct fuse_compound_out *compound_out;
+	size_t offset;
+	int i;
+
+	if (!compound || !compound->response_data)
+		return -EINVAL;
+
+	/* double parsing will be important
+	 * for large responses most likely out pages.
+	 */
+	if (compound->parsed)
+		return 0;
+
+	if (compound->response_size < sizeof(struct fuse_compound_out))
+		return -EINVAL;
+
+	compound_out = (struct fuse_compound_out *)compound->response_data;
+	compound->parsed_header = compound_out;
+
+	offset = sizeof(struct fuse_compound_out);
+
+	/* Parse each operation response and only look at the arguments we
+	 * have responses for
+	 */
+	for (i = 0; i < compound_out->count && i < compound->count; i++) {
+		struct fuse_out_header *hdr;
+		struct fuse_args *args = compound->op_args[i];
+		size_t data_size;
+		void *response_data;
+		int arg_idx;
+
+		if (offset + sizeof(struct fuse_out_header) >
+		    compound->response_size)
+			return -EINVAL;
+
+		hdr = (struct fuse_out_header *)
+			((char *)compound->response_data + offset);
+
+		if (hdr->len < sizeof(struct fuse_out_header))
+			return -EINVAL;
+
+		if (offset + hdr->len > compound->response_size)
+			return -EINVAL;
+
+		data_size = hdr->len - sizeof(struct fuse_out_header);
+		response_data = (char *)compound->response_data + offset +
+				sizeof(struct fuse_out_header);
+
+		/* Store error code */
+		compound->op_errors[i] = hdr->error;
+
+		/* Copy response data on success */
+		if (!hdr->error && args && args->out_numargs > 0 && data_size > 0) {
```

Comment:
Can we just stop the loop and remove indentation levels? Like

```
if (hdr->error)
   break;
```

Reply:
Or 

```
		if (!hdr->error && args && args->out_numargs > 0 && data_size > 0)
		   break;
```

We could use `continue`, but it is a bit weird to continue if there is an error.